src/DependencyGraph/ResolutionRequest.js

```diff
@@ -127,79 +129,122 @@ class ResolutionRequest {
-                name,
-                mod.path
-              );
-              return false;
-            }
-            return filteredPairs.push([name, modDep]);
+            });
+            return [dependencyNames, dependencies];
           });
+        };
+
+      const collectedDependencies = defaultMap(module => collect(module));
+      const crawlDependencies = (mod, [depNames, dependencies]) => {
+        const filteredPairs = [];
+
+        dependencies.forEach((modDep, i) => {
+          const name = depNames[i];
+          if (modDep == null) {
+            // It is possible to require mocks that don't have a real
+            // module backing them. If a dependency cannot be found but there
+            // exists a mock with the desired ID, resolve it and add it as
+            // a dependency.
+            if (allMocks && allMocks[name] && !mocks[name]) {
+              const mockModule = this._moduleCache.getModule(allMocks[name]);
+              mocks[name] = allMocks[name];
+              return filteredPairs.push([name, mockModule]);
+            }
 
-          response.setResolvedDependencyPairs(mod, filteredPairs);
-
-          const newDependencies =
-            filteredPairs.filter(([, modDep]) => !visited[modDep.hash()]);
-
-          if (onProgress) {
-            finishedModules += 1;
-            totalModules += newDependencies.length;
-            onProgress(finishedModules, totalModules);
+            debug(
+              'WARNING: Cannot find required module `%s` from module `%s`',
+              name,
+              mod.path
+            );
+            return false;
           }
-          return Promise.all(
-            newDependencies.map(([depName, modDep]) => {
-              visited[modDep.hash()] = true;
-              return Promise.all([modDep, recursive ? collect(modDep) : []]);
-            })
-          );
+          return filteredPairs.push([name, modDep]);
         });
+
+        response.setResolvedDependencyPairs(mod, filteredPairs);
+
+        const dependencyModules = filteredPairs.map(([, m]) => m);
+        const newDependencies =
+          dependencyModules.filter(m => !collectedDependencies.has(m));
+
+        if (onProgress) {
+          finishedModules += 1;
+          totalModules += newDependencies.length;
+          onProgress(finishedModules, totalModules);
+        }
+
+        if (recursive) {
+          // doesn't block the return of this function invocation, but defers
+          // the resulution of collectionsInProgress.emptied.then(…)
+          dependencyModules
+            .forEach(dependency => collectedDependencies.get(dependency));
+        }
+        return dependencyModules;
       };
 
-      return collect(entry).then(deps => {
-        recursiveFlatten(deps).forEach(dep => response.pushDependency(dep));
+      const collectionsInProgress = emptyingSet();
+      function collect(module) {
```

Comment:
nit: should this function be defined above with the other ones before it's invoked for clarity?